File: _Uncategorized/232_Stack_ImplementQueue.cs
// 232. Implement Queue using Stacks
// Tags: quirky, stack
//

public class MyQueue 
{
    private Stack<int> ins;
    private Stack<int> outs;
    
    public MyQueue() 
    {
        ins = new();
        outs = new();
    }
    
    public void Push(int x) 
    {
        ins.Push(x);
    }
    
    public int Pop() 
    {
        this.Peek();
        return outs.Pop();
    }
    
    public int Peek() 
    {
        if(outs.Count == 0)
        {
            while(ins.Count > 0)
            {
                outs.Push(ins.Pop());
            }
        }
        return outs.Peek();
    }
    
    public bool Empty() 
    {
        if(ins.Count == 0 && outs.Count == 0)
            return true;
        else
            return false;
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.Push(x);
 * int param_2 = obj.Pop();
 * int param_3 = obj.Peek();
 * bool param_4 = obj.Empty();
 */

File: _Uncategorized/232_stack_ImplementQueue.rs
// 232. Implement Queue using Stacks
// Tags: quirky, stack
//
/**
 * Your MyQueue object will be instantiated and called as such:
 * let obj = MyQueue::new();
 * obj.push(x);
 * let ret_2: i32 = obj.pop();
 * let ret_3: i32 = obj.peek();
 * let ret_4: bool = obj.empty();
 */

struct MyQueue {
    ins: Vec<i32>,
    outs: Vec<i32>,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
 impl MyQueue {

    fn new() -> Self {
        return MyQueue { ins: Vec::new(), outs: Vec::new(), };
    }
    
    fn push(&mut self, x: i32) {
        self.ins.push(x);
    }
    
    fn pop(&mut self) -> i32 {
        self.peek();

        match self.outs.pop() {
            None => -1,
            Some(n) => (*n),
        }
    }
    
    // Populates the outs stack and returns the top of the
    // outs stack.
    fn peek(&mut self) -> i32 {
        if self.outs.len() == 0 {
            while self.ins.len() > 0 {

                match self.ins.pop() {
                    None => self.outs.push(0),
                    Some(n) => self.outs.push(n),    
                }
            }
        }

        // Return the last element or 0 if it doesn't exist.
        match self.outs.last() {
            None => 0,
            Some(n) => n
        }
    }
    
    fn empty(&self) -> bool {
        if self.ins.is_empty() && self.outs.is_empty() {
            return true;
        }

        false
    }
}